mozetl/taar/taar_locale.py

```diff
@@ -0,0 +1,177 @@
+          AND value['app_disabled'] = FALSE -- exclude compatibility disabled addons
+          AND value['is_system'] = FALSE -- exclude system addons
+          AND locality <> 'null'
+          AND key is not null
+        ),
+
+        country_addon_pairs AS (
+        SELECT
+        COUNT(*) AS pair_cnts, addon_key, locality
+        from filtered_sample
+        GROUP BY locality, addon_key
+        )
+
+        SELECT * FROM country_addon_pairs
+        ORDER BY locality, pair_cnts DESC
+    """)
+
+
+def compute_threshold(addon_df):
+    """ Get a threshold to remove locales with a small
+    number of addons installations.
+    """
+    addon_install_counts = addon_df.groupBy('locality')\
+        .agg({'pair_cnts': 'sum'})\
+        .rdd.map(lambda p: p[1])\
+        .collect()
+
+    # Compute a threshold at the 25th percentile to remove locales with a
+    # small number of addons installations.
+    locale_pop_threshold = np.percentile(np.array(addon_install_counts), 25)
+
+    # Safety net in case the distribution gets really skewed, we should
+    # require 2000 addon installation instances to make recommendations.
+    return 2000 if locale_pop_threshold < 2000 else locale_pop_threshold
+
+
+def transform(addon_df, threshold, num_addons):
+    """ Converts the locale-specific addon data in to a dictionary.
+
+    :param addon_df: the locale-specific addon dataframe;
+    :param threshold: the minimum number of addon-installs per locale;
+    :param num_addons: requested number of recommendations.
+    :return: a dictionary {<locale>: ['GUID1', 'GUID2', ...]}
+    """
+    top10_per = {}
+
+    # Decide that we can not make reasonable recommendations without
+    # a minimum number of addon installations.
+    grouped_addons = addon_df\
+        .groupBy('locality')\
+        .agg({'pair_cnts': 'sum'})\
+        .collect()
+    list_of_locales =\
+        [i['locality'] for i in grouped_addons if i['sum(pair_cnts)'] > threshold]
+
+    for specific_locale in list_of_locales:
+        # Most popular addons per locale sorted by number of installs
+        # are added to the list.
+        sorted_addon_guids = addon_df\
+          .filter(addon_df.locality == specific_locale)\
+          .sort(addon_df.pair_cnts.desc())\
+          .collect()
+
+        # Creates a dictionary of locales (keys) and list of
+        # recommendation GUIDS (values).
+        top10_per[specific_locale] =\
+            [addon_stats.addon_key for addon_stats in sorted_addon_guids[0:num_addons]]
+
+    return top10_per
+
+
+def store(dictionary, prefix, bucket):
+    FULL_FILENAME = "{}.json".format(LOCALE_FILE_NAME)
+
+    with open(FULL_FILENAME, "w+") as dict_file:
+        # If we attempt to load invalid JSON from the assembled file,
+        # the next function throws
+        dict_file.write(json.dumps(dictionary, indent=2))
+
+    archived_file_copy =\
+        "{}{}.json".format(LOCALE_FILE_NAME, datetime.today().strftime("%Y%d%m"))
```

Comment:
The date should be taken as an argument from the command line, to ensure idempotency. This allows for things like airflow backfill.

If you are not running on airflow, you could run this script manually (or scheduled via ATMO) through a script in this form.
```
# In[ ]:

# Stop the injected spark session from interfering with mozetl
spark.stop()

# In[ ]:

%%bash
cd /tmp
wget https://raw.githubusercontent.com/mozilla/python_mozetl/master/bin/mozetl-submit.sh
bash mozetl-submit.sh \
    -p https://github.com/mozilla/python_mozetl.git \
    taar_locale --start_date $(date -d '-1 day' '+%Y%m%d') \
        --foo value_1 \
        --bar value_2 
```

Reply:
Mh, good point. We just want the latest data for this job and don't care about older data (we just save a copy for future reference, but we're not planning to use it).

With that said, is ATMO scheduling preferred over Airflow scheduling in this case?

Reply:
No, you should probably use airflow (email notifications, retries). Mozetl in airflow essentially runs the second cell, but has a wrapper that passes context via environment variables instead of command line arguments.

If you're going to embed the date into the filename, I would still suggest taking the date from the environment. Ideally an ETL job should be able to run multiple times with the same date and always end up in the same state.

Reply:
Done :)